### DISCUSSION TITLE
Polygon Clipping

### DISCUSSION THREAD

**Discussion by kurt-rhee** [dotnet/Silk.NET#1338] [Q&A]
Hello, 

I am working on a project which would implement two different algorithms and we are currently trying to decide what library if any might help us achieve fast behavior.  I attempted to search the issues and discussion boards for the terms "polygon clipping," but could not find related answers, if I missed anything please feel free to point me in the right direction.

The first is a shadow display tool in the web.  We want the user to be able to create objects (convex polygons numbering in the millions) in 3 dimensions and for the tool to render the shade from the sun (as a directional light source) at a given timestamp throughout the year.  This rendering should be indicative, but does not need to be mathematically rigorous.  My thought was shadow mapping might be a good use case for this, and my understanding is that silk.net does not support the browser.  Is this true?

The second use case is the actual calculation engine for the 3d scene mentioned above.  This calculation engine does need to be mathematically rigorous, but does not necessarily need to render to the user unless it was faster than shadow mapping.  My thought was to project objects into a plane perpendicular to the directional light source and then do polygon clipping operations.  Is this possible in Silk.net?  If so would the polygon projects be done on the GPU and would the polygon clipping operations also be done on the GPU?

Thank you for your help!

**Comment by HurricanKai** [comment]
Hello,

Silk.NET simply provides access to low-level APIs, which in turn provide GPU access. @Beyley is working on some Web support, they can comment on the status of that.

Other then that, I'm confident that what you describe can be implemented with heavy use of the GPU to massively speed up calculation. Wondering what "mathematically rigorous" means here, depending on that you might be interested in various shadow techniques from live mediums (like games) or from pre-rendering production like movies. In recent times ray-tracing has been widely adopted too, providing impressive results, but depending on what sort of users you need to be available to this might not be an option. All of these options are not made to accurately predict the real world, they are made to impress and fool a human viewer (which makes them fairly accurate, but not perfect).

True physics simulations are often using only compute abilities, which as you mention aren't aimed at rendering much to a user. To give some advice here I'd have to know what the intended result here is.

Regarding projection & clipping, this is just modern rendering. What you refer to as clipping is I believe what is commonly referred to as rasterizing (deciding whether a polygon intersects / overlaps with a pixel of the resulting image). Projecting is trivial, it's just some basic matrix maths. This operation can be implemented trivially with any rendering API, which will provide good speed. The calculation of overlapping fragments is done by dedicated hardware, but can be slightly configured to do what you need.

Very happy to help :slightly_smiling_face:

**Reply by kurt-rhee** [reply]
Hello thanks so much for this reply, it is very helpful.  I should have been more specific in my original question regarding the polygon "clipping" algorithm.  My domain is in the field of predicting the amount of energy a solar power plant will produce and not in graphics processing.  I am quite the newbie when it comes to the terminology for this field.

What I would like to do is to take a set of polygons (solar panels), project them into a plane perpendicular to the incoming directional light rays (rays from the sun) and then perform polygon-polygon intersections to determine the shape of the shadow on each individual polygon.  We can then use the resulting shadow shape to predict the solar panel's performance.  What I originally meant by mathematically rigorous would probably be better described by geometrically rigorous?  I have written this program using Python in the past, but with the number of polygons in my scenes and the number of scenes the results would often take hours.  My hope is to rewrite the algorithm using C sharp and process as much as I can on the GPU instead of the CPU.  

Basically we cannot use an algorithm which is meant to fool a human being, or anything which takes into account pixels, mostly because we must be able to prove to a bank and accompanying set of independent engineers that our shade engine is geometrically accurate.  

Based on your answer above, it seems as though it is possible to take a set of polygons and project them using the GPU.  Is it possible to also calculate polygon-polygon intersections in 2D on the GPU and then store the resulting shade shape?  I was not sure what the term "dedicated hardware" means.
